src/qp_klp/Assays.py

```diff
@@ -144,6 +143,141 @@ def _generate_artifact_name(self, prep_file_path):
+            self.generate_prep_file()
+
+        # moved final component of genprepfilejob outside of object.
+        # obtain the paths to the prep-files generated by GenPrepFileJob
+        # w/out having to recover full state.
+        tmp = join(self.pipeline.output_path, 'GenPrepFileJob', 'PrepFiles')
+
+        self.has_replicates = False
+
+        prep_paths = []
+        self.prep_file_paths = {}
+
+        for root, dirs, files in walk(tmp):
+            for _file in files:
+                # we are looing for .tsv files and we are only interested
+                # in the string after the last _, which is the study_id
+                if not _file.endswith('.tsv'):
+                    continue
+                # continue if no underscore
+                chunks = _file.rsplit('_', 1)
+                if len(chunks) <= 1:
+                    continue
+                # continue if no int after .
+                qid = chunks[-1].split('.')[0]
+                if not qid.isnumeric():
+                    continue
+                if qid not in self.prep_file_paths:
+                    self.prep_file_paths[qid] = []
+
+                _path = abspath(join(root, _file))
+                prep_paths.append(_path)
+                self.prep_file_paths[qid].append(_path)
+
+            for _dir in dirs:
+                if _dir == '1':
+                    # if PrepFiles contains the '1' directory, then it's a
+                    # given that this sample-sheet contains replicates.
+                    self.has_replicates = True
+
+        # currently imported from Assay although it is a base method. it
+        # could be imported into Workflows potentially, since it is a post-
+        # processing step. All pairings of assay and instrument type need to
+        # generate prep-info files in the same format.
+        self.overwrite_prep_files(prep_paths)
+
+        # for now, simply re-run any line below as if it was a new job, even
+        # for a restart. functionality is idempotent, except for the
+        # registration of new preps in Qiita. These will simply be removed
+        # manually.
+
+        # post-processing steps are by default associated with the Workflow
+        # class, since they deal with fastq files and Qiita, and don't depend
+        # on assay or instrument type.
+        self.update_status("Generating sample information", 5, 9)
+        self.sifs = self.generate_sifs()
+
+        # post-processing step.
+        self.update_status("Registering blanks in Qiita", 6, 9)
+        if self.update:
+            self.update_blanks_in_qiita()
+
+        self.update_status("Loading preps into Qiita", 7, 9)
+        if self.update:
+            self.update_prep_templates()
+
+        # before we load preps into Qiita we need to copy the fastq
+        # files n times for n preps and correct the file-paths each
+        # prep is pointing to.
+        self.load_preps_into_qiita()
+
+        self.fsr.generate_report()
```

Comment:
Could we toss this line a comment, too? I see that it was in `MetaOmic.execute_pipeline` but not in `StandardAmpliconWorkflow.execute_pipeline`, so now I'm curious about why ... :D. I also notice that at one point in `MetaOmic` the code checks `if hasattr(self, 'fsr'):` before calling `self.fsr.<something>()` ... do we need to worry that not every assay type or every individual assay instance will have an `fsr` property?

Reply:
fsr is an instance of FailedSamplesRecord, which is its own object, and as far as I can tell is used with Job.audit to keep track of the samples lost on each of the steps of the pipeline. Here we call FailedSamplesRecord. generate_report to output the report so it's moved to the final output and the user can get access. Adding this info in the code.